[ROCProfiler-SDK] Remove Python library dependency from Python bindings

ammarwa opened the pull request:
## Motivation

Modified CMake configuration to build Python extension modules without linking against libpython, enabling compatibility with manylinux containers and Python installations that don't provide shared Python libraries.

In ManyLinux_2.28, if `-DROCPROFILER_PYTHON_VERSIONS="3.8;3.9;3.10;3.11;3.12;3.13;3.14"`is used, then we will have rocpd and roctx compatible with Python version 3.8 or higher.

This solution addresses [Issue #1808](https://github.com/ROCm/TheRock/issues/1808)

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details

- Changed `FindPython3` component from "Development" to "Development.Module" to only require Python headers, not the shared library
- Added a fallback to the old way of finding the Python3 Development component for sanitizers as well.
- Removed `${Python3_LIBRARIES}` from `target_link_libraries` for both `rocprofiler-sdk-roctx-python-bindings` and `rocprofiler-sdk-rocpd-python-bindings` in case if sanitizers are not enabled
- Python symbols are now resolved at runtime from the interpreter rather than linked at build time
- General Improvements to packaging 

Benefits:
  - Python bindings can now be built in manylinux containers without installing `python3-dev/devel` packages for each Python version
  - Bindings work with Python installations that lack `libpython.so` (standard in manylinux base images)
  - Reduced build dependencies while maintaining full functionality

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan

- Tested manually using `quay.io/pypa/manylinux_2_28_x86_64` docker image, which is similar to what is used in the Rock CI system, I used the following cmake configuration: 

```bash
cmake -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_INSTALL_PREFIX=$(realpath /opt/rocm) -DROCPROFILER_BUILD_{CI,SAMPLES,TESTS}=ON -DPython3_EXECUTABLE=$(which python3.8) -DROCPROFILER_PYTHON_VERSIONS="3.8;3.9;3.10;3.11;3.12;3.13;3.14" -DCPACK_GENERATOR='DEB;RPM;TGZ' -DCPACK_PACKAGING_INSTALL_PREFIX="$(realpath /opt/rocm)" .
```

- The Rock CI System will be able to make sure that this is being packaged and produced correctly every time.

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

- Verified no libpython in ldd output for built modules
- Confirmed modules load successfully with Python 3.8
- DEB, RPM, and TGZ packages built successfully
- Tests successfully passed locally

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [X] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

ammarwa committed:
Remove Python library dependency from Python bindings for manylinux compatibility

ammarwa committed:
Fixing Sanitization Issues

ammarwa commented:
<details>
<summary><h2>Code Coverage Report</h2></summary>

<!-- Keep the empty line above to render markdown properly -->
# Code Coverage Report

## Tests Only

![code coverage tests.png](https://github.com/ROCm/rocm-systems/blob/3c089191175d026855050960c7d995dc9f4ad391/pr-1451/tests.png?raw=True)

## Samples Only

![code coverage samples.png](https://github.com/ROCm/rocm-systems/blob/3c089191175d026855050960c7d995dc9f4ad391/pr-1451/samples.png?raw=True)

## Tests + Samples

![code coverage all.png](https://github.com/ROCm/rocm-systems/blob/3c089191175d026855050960c7d995dc9f4ad391/pr-1451/all.png?raw=True)

<!-- code-coverage-comment-identifier -->
</details>

jrmadsen commented on projects/rocprofiler-sdk/source/lib/python/utilities.cmake:
```diff
@@ -65,7 +82,11 @@ function(get_default_python_versions _VAR)
     # If none found, do one last check for 3.6 (no EXACT)
     if(NOT _PYTHON_FOUND_VERSIONS)
-        find_package(Python3 3.6 COMPONENTS Interpreter Development)
+        if(ROCPROFILER_MEMCHECK OR CMAKE_VERSION VERSION_LESS "3.18")
```
https://github.com/ROCm/rocm-systems/blob/61fc256db9ecd3bb5b0c0b4b537b48caae085f65/projects/rocprofiler-sdk/CMakeLists.txt#L1

ammarwa committed:
Refactor comments in rocprofiler_find_python3 macro

Removed redundant comments and improved readability.

ammarwa committed:
Refactor Python package finding logic

jrmadsen commented on projects/rocprofiler-sdk/cmake/rocprofiler_config_packaging.cmake:
```diff
@@ -194,6 +202,9 @@ set(CPACK_RPM_PACKAGE_AUTOREQ
 set(CPACK_RPM_TESTS_PACKAGE_AUTOPROV OFF) # disable for tests package
+# Python extension modules should not have automatic dependency generation
+set(CPACK_RPM_ROCPD_PACKAGE_AUTOPROV OFF) # Python modules don't provide shared libraries
+set(CPACK_RPM_ROCTX_PACKAGE_AUTOPROV OFF) # Python modules don't provide shared libraries
```
```suggestion
```

The rocpd and roctx packages include more than the Python modules. They include the shared libraries `librocprofiler-sdk-rocpd.so` and `librocprofiler-sdk-roctx.so`.

jrmadsen commented on projects/rocprofiler-sdk/cmake/rocprofiler_config_packaging.cmake:
```diff
@@ -160,6 +160,14 @@ set(CPACK_DEBIAN_PACKAGE_GENERATE_SHLIBS ON) # generate list of shared libs prov
+set(CPACK_DEBIAN_ROCPD_PACKAGE_GENERATE_SHLIBS OFF) # Python modules don't need SONAME
+set(CPACK_DEBIAN_ROCTX_PACKAGE_SHLIBDEPS OFF) # Python modules don't need dependency
+                                              # scanning
+set(CPACK_DEBIAN_ROCTX_PACKAGE_GENERATE_SHLIBS OFF) # Python modules don't need SONAME
```
```suggestion
```
The rocpd and roctx packages include `librocprofiler-sdk-rocpd.so` and `librocprofiler-sdk-roctx.so`

jrmadsen committed:
Misc updates

- fix package configuration
- fix git revision / git describe

jrmadsen committed:
Tweak python/utilities.cmake

- use variable instead of if/else for ROCPROFILER_MEMCHECK

jrmadsen committed:
Fix linking to Python3_LIBRARIES

- replace if(ROCPROFILER_MEMCHECK) with if("Development" IN_LIST ROCPROFILER_BUILD_Find_Python3_COMPONENTS)

jrmadsen committed:
Fix comments
